### PR TITLE
feat(protocol): implement & simulate tokenomics

### LINKED ISSUE #26: feat(protocol): design and implement tokenomics v1 for review
Some very useful discussions are available in [this issue](https://github.com/taikoxyz/taiko-mono/issues/14)

### PR DESCRIPTION
**Pull request by dantaik**
This new PR extracts tokenomics-only code changes so it will be much easier to review now.
It also replace the previous LibConstants with a new config object that can be overridden to make tests easier.

### CODE REVIEW

**Comment by dantaik** [comment]
@cyberhorsey @davidtaikocha shall we merge the PR? The feature by default, is disabled.

File: packages/protocol/contracts/L1/libs/LibProposing.sol [hunk 1]
```diff
@@ -58,33 +61,38 @@ library V1Proposing {
...
+        });
+        _validateMetadata(config, meta);
 
         {
             bytes calldata txList = inputs[1];
             // perform validation and populate some fields
             require(
-                txList.length >= 0 &&
-                    txList.length <= LibConstants.K_TXLIST_MAX_BYTES &&
+                txList.length > 0 &&
```
Comment (davidtaikocha) [inline packages/protocol/contracts/L1/libs/LibProposing.sol:88]: Seems we will not allow proposing empty block by giving empty encoded txlist bytes? Is this change intended?

**Review by davidtaikocha — DISMISSED** [review]
lets merge this now? looks like there is only one feedback unresloved.

**Comment by RogerLamTd** [comment]
> lets merge this now? looks like there is only one feedback unresloved.

should i approve for merge then?

File: packages/protocol/contracts/L1/libs/LibProposing.sol (hunk 1 above)
Comment (dantaik) [reply packages/protocol/contracts/L1/libs/LibProposing.sol]: Good catch, fixed.

**Review by davidtaikocha — DISMISSED** [review]

**Review by davidtaikocha — DISMISSED** [review]

**Review by cyberhorsey — APPROVED** [review]

**Review by davidtaikocha — APPROVED** [review]
